phalcon/Mvc/Collection.zep

```diff
@@ -32,7 +31,7 @@ use Serializable;
  * This component implements a high level abstraction for NoSQL databases which
  * works with documents
  */
-abstract class Collection implements EntityInterface, CollectionInterface, InjectionAwareInterface, Serializable
+abstract class Collection implements EntityInterface, CollectionInterface, InjectionAwareInterface, Unserializable, \Serializable
```

Comment:
Why do we need implement Unserializable? Can we avoid this?

Reply:
It is this interface and the bsonUnserialize method that allows you to fill in the result sets

Reply:
Not all users have mongodb installed. Declaring interface here will cause fatal error for those users who don not have mongodb installed even they not use Collection class at all. Just at PHP startup time

Reply:
If I delete this interface, MongoDB raises an Exception

`[MongoDB\Driver\Exception\InvalidArgumentException] Class Phalcon\Test\Fixtures\Mvc\Collections\Robots does not implement MongoDB\BSON\Unserializable`

https://www.php.net/manual/en/mongodb-driver-cursor.toarray.php

Reply:
Can you declare Robots implemets Unserializable and leave Collection as is?

Reply:
@

> If I delete this interface, MongoDB raises an Exception
> 
> `[MongoDB\Driver\Exception\InvalidArgumentException] Class Phalcon\Test\Fixtures\Mvc\Collections\Robots does not implement MongoDB\BSON\Unserializable`
> 
> https://www.php.net/manual/en/mongodb-driver-cursor.toarray.php

We cannot use that interface there because that introduces a limitation for the framework. Here is why:

We have introduced PDO (a few versions back). That came with some interfaces that we wanted to use to wrap Phalcon classes. As a result, because these are interfaces, we added the requirement to load PDO first before Phalcon

With v4 we introduced PSR-7 and its interfaces. New requirement: PSR has to be loaded first.

If we add this interface there, it will require all users, irrespective of whether they want to use Mongo or not, to load the mongo extension for Phalcon to work.

You might ask why does `apcu` work then if I do not have it installed? It works because we use functions for it and not interfaces. Another example is one of the `Redis` adapters. If you check the code, we create the adapter inside Phalcon and use it from there, but never expose any interfaces that Redis or any other extension uses.

I think the best way forward is to remove the interface from there and require models to implement the `Unserializable` interface for the mongo collection to work.

In short, if we introduce one interface from a different extension, Zend Engine requires it in every request, even though you might not want to use that interface ever.

Reply:
The solution is what then, to create an interface if it does not exist, or not to use the "typeMap" functionality ?

Reply:
https://www.php.net/manual/fr/mongodb.persistence.deserialization.php
This method still offers many possibilities. It allows you to transform a field into a class, for example

Reply:
Require models to implement the Unserializable interface for the MongoDB collection will do all we need here. Just  implement all interface requirements in this PR but do not declare "implement ..." interface. Indicate in the comment that the user must declare such a inheritance for each of its model of his own.

Reply:
Okay, I'll fix it and take care of it.